### DISCUSSION TITLE
Webpack Module Federation: `sharedMappings` causing "Call Stack Size Exceeded" error in Safari

### DISCUSSION THREAD

**Discussion by Kamble-Prafful** [webpack/webpack#18776] [Q&A]
I've been debugging my Angular Micro-Frontend Nx Monorepo project to resolve an issue where the web app shows a blank page on iOS (Safari & Chrome) and macOS (Safari). The app works perfectly fine on Chrome on macOS.

### **Issue**:

- The **shell app** (host) shows **no console errors** on Safari.
- The **remote apps** show **RangeError: Maximum call stack size exceeded**.

After some debugging, I was finally able to find the root cause of the error in **webpack.config.js**. We are using **sharedMappings** from `'@angular-architects/module-federation/webpack'` to share a custom library (`@project-name/web/shared/util`). When I comment out the shared lib configuration in `sharedMappings()`, the remote apps start working fine in Safari, but the shell app breaks when trying to use the shared library.

### **Hypothesis**:

We have three apps:

1. **Shell (host)** app.
2. **Remote App 1**.
3. **Remote App 2**.

These apps share some common services and utilities (such as authentication and configuration services) that need to be singleton across apps when I'm using Shell App. when the shell accesses resources from a remote app, the remote app also has its shared lib's instance but the shell app's shared lib instance should be used here. So that's why we mention the shared lib's path while registering sharedMappings() in the Webpack file.

So correct me if it doesn't work like this.

### Workarounds Tried

I've tried many approaches before identifying the root cause in the Webpack configuration.

- After finding the issue in the Webpack file, I attempted to configure it differently by placing the custom library directly in `shared: share({ ... })` and specifying `eager: true`. Unfortunately, this didn't resolve the issue.

- I also worked on detecting and resolving circular dependencies throughout the entire project, especially in the shared library, using various tools such as **madge**, **dpdm**, and **circular-dependency-plugin**. However, the issue persisted.

Nothing seems to be working as of now. It's working fine in windows for all browsers such as edge, chrome, etc. But not for iOS and macOS' safari. I'm not sure if the problem is in the Webpack configuration, the way we're sharing the library, or if there’s some issue with the custom library itself when running in Safari. 

I’ve provided the relevant portion of my Webpack config below.

### References:
**webpack.config.js**
```js
const ModuleFederationPlugin = require('webpack/lib/container/ModuleFederationPlugin');
const mf = require('@angular-architects/module-federation/webpack');
const path = require('path');
const share = mf.share;

/**
 * We use the NX_TSCONFIG_PATH environment variable when using the @nrwl/angular:webpack-browser
 * builder as it will generate a temporary tsconfig file which contains any required remappings of
 * shared libraries.
 * A remapping will occur when a library is buildable, as webpack needs to know the location of the
 * built files for the buildable library.
 * This NX_TSCONFIG_PATH environment variable is set by the @nrwl/angular:webpack-browser and it contains
 * the location of the generated temporary tsconfig file.
 */
const tsConfigPath = process.env.NX_TSCONFIG_PATH ?? path.join(__dirname, '../../tsconfig.base.json');

const workspaceRootPath = path.join(__dirname, '../../');
const sharedMappings = new mf.SharedMappings();
sharedMappings.register(
	tsConfigPath,
	[
		/* mapped paths to share */
		'@project-name/web/shared/util'
	],
	workspaceRootPath
);

const featureModulesPath = {
'./DbUpgrade': 'libs/web/common/db-upgrade/feature/src/lib/db-upgrade-feature.module.ts',
...and many more
};

module.exports = {
	experiments: {
		outputModule: true
	},
	output: {
		uniqueName: 'common_mfe',
		publicPath: 'auto'
	},
	optimization: {
		runtimeChunk: false,
		minimize: true
	},
	resolve: {
		alias: {
			...sharedMappings.getAliases()
		}
	},
	plugins: [
		new ModuleFederationPlugin({
			library: {
				type: 'module'
			},
			name: 'common_mfe',
			filename: 'remoteEntry.js',
			exposes: featureModulesPath,
			shared: share({
				'@angular/core': { singleton: true, strictVersion: true, requiredVersion: 'auto' },
				'@angular/forms': { singleton: true, strictVersion: true, requiredVersion: 'auto' },
				'@angular/common': { singleton: true, strictVersion: true, requiredVersion: 'auto' },
				'@angular/common/http': { singleton: true, strictVersion: true, requiredVersion: 'auto' },
				'@angular/animations': { singleton: true, strictVersion: true, requiredVersion: 'auto' },
				'@angular/router': { singleton: true, strictVersion: true, requiredVersion: 'auto' },
				rxjs: { singleton: true, strictVersion: true, requiredVersion: 'auto', includeSecondaries: true },
				'rxjs/operators': { singleton: true, strictVersion: true, requiredVersion: 'auto' },
				'@infragistics/igniteui-angular': {
					singleton: true,
					strictVersion: true,
					requiredVersion: 'auto',
					includeSecondaries: true
				},
				'lodash-es': { singleton: true, strictVersion: true, requiredVersion: 'auto', includeSecondaries: true },
				primeng: { singleton: true, strictVersion: true, requiredVersion: 'auto', includeSecondaries: true },
				'xng-breadcrumb': { singleton: true, strictVersion: true, requiredVersion: 'auto' },
				...sharedMappings.getDescriptors()
			})
		}),
		sharedMappings.getPlugin()
	]
};
```

**Comment by alexander-akait** [comment]
Sounds like you have circular deps, better to debug code and show stacktrace

**Reply by Kamble-Prafful** [reply]
I've already tried this by detecting and resolving all the circular dependencies using tools such as madge, dpdm, circular-dependency-plugin, etc in my whole project. But still, it's showing me the same error.

**Reply by alexander-akait** [reply]
please create reproducible test repo, without code I can't help
